### PR TITLE
python3Packages.perfplot: 0.5.0 -> 0.8.4, fix build by packaging new dependencies

### PR DESCRIPTION
**Pull request by risicle**
###### Motivation for this change
ZHF: #97479

Add new python packages `dufte` and `rich` to satisfy `perfplot` 0.8.4's dependencies.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/rich/default.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  # tests not included in pypi tarball
+  src = fetchFromGitHub {
+    owner = "willmcgugan";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "0hv27b22x7dbx1i7nzsd8y8fymmvdak2hcx9242jwk4c1a7jr151";
+  };
+  format = "pyproject";
+
+  buildInputs = [ poetry ];
```
Comment (jonringer) [inline pkgs/development/python-modules/rich/default.nix:28]: generally poetry is just used for packaging
```suggestion
  nativeBuildInputs = [ poetry ];
```

File: pkgs/development/python-modules/dufte/default.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    inherit pname version;
+    sha256 = "0nkaczipbsm8c14j9svxry2wigmn5iharibb6b8g062sjaph8x17";
+  };
+  format = "pyproject";
+
+  propagatedBuildInputs = [
+    importlib-metadata
+    matplotlib
+    numpy
+  ];
```
Comment (jonringer) [inline pkgs/development/python-modules/dufte/default.nix:26]: https://github.com/nschloe/dufte/blob/b08815ea12c6dbff419ff65aa36860f2af54afaf/setup.cfg#L30
```suggestion
  propagatedBuildInputs = [
    matplotlib
    numpy
  ] ++ lib.optionals (pythonOlder "3.8") [
    importlib-metadata
  ];
```

Comment (risicle) [reply pkgs/development/python-modules/dufte/default.nix]: Did we ever consider making these "backport" libraries simply evaluate to null (or something else harmless/empty) on a non-applicable python version?

**Review by drewrisinger — APPROVED** [review]
* [x] Diff LGTM
* [x] Commits LGTM
* [x] Builds via ``nix-review`` on linux-86_64:

```
https://github.com/NixOS/nixpkgs/pull/100168
6 packages built:
python37Packages.dufte python37Packages.perfplot python37Packages.rich python38Packages.dufte python38Packages.perfplot python38Packages.rich
```

File: pkgs/development/python-modules/dufte/default.nix (hunk 2 above)
Comment (jonringer) [reply pkgs/development/python-modules/dufte/default.nix]: a lot of them do
```
[10:13:08] jon@nixos ~/projects/nixpkgs/pkgs/development/python-modules (backport-python-dask)
$ rg "then null"
enum34/default.nix
8:if pythonAtLeast "3.4" then null else buildPythonPackage rec {

ipaddress/default.nix
8:if (pythonAtLeast "3.3") then null else buildPythonPackage rec {

backports_shutil_get_terminal_size/default.nix
8:if !(pythonOlder "3.3") then null else buildPythonPackage {

cffi/default.nix
3:if isPyPy then null else buildPythonPackage rec {

functools32/default.nix
7:if isPy3k then null else buildPythonPackage rec {

backports_lzma/default.nix
10:if !(pythonOlder "3.3") then null else buildPythonPackage rec {
```
but it still causes failure for some https://github.com/NixOS/nixpkgs/pull/100017

Comment (jonringer) [reply pkgs/development/python-modules/dufte/default.nix]: see https://github.com/rytilahti/python-miio/pull/828 as an example

**Pull request merged**